### DISCUSSION TITLE
Proposal: Disallow non-vocabulary keywords

### DISCUSSION THREAD

**Discussion by handrews** [json-schema-org/community#241] [Specification]
This is based on numerous recent discussions including @awwright raising it on a recent community call, @karenetheridge raising it in an [issue comment](https://github.com/json-schema-org/JSON-Schema-Test-Suite/issues/574#issuecomment-1229547081), and possibly others I'm forgetting, with apologies.

For further discussion on how to continue support for this, please see https://github.com/orgs/json-schema-org/discussions/329.

## Terminology

* a keyword is _supported_ if the implementation understands to run a specific piece of code for it
* a keyword is _understood_ if the implementation knows what vocabulary it comes from, but doesn't have specific code for it
* a keyword is _unknown_ if the implementation cannot figure out whether it comes from any particular vocabulary or not
* a _simple value annotation_ (_SVA_) keyword is one that unconditionally uses the exact keyword value as the annotation value

Therefore, `title` and `readOnly` are SVAs.  Hyper-Schema's `links` is not, because it is a template that is filled out with instance data, and therefore requires custom code to support properly.

## Scope of this discussion

This topic interacts with many complex topics, but let's try to keep this discussion by assuming that certain problems are solvable.  It is fair game to raise doubts over that solvability since the proposal here requires that these problems are solved, but let's not dive into _how_ to solve these things here (I have tons of thoughts on several of these points, and I know others do as well).

**Edit** - A new discussion on _how_ to solve these things is now [here](https://github.com/orgs/json-schema-org/discussions/329).

### Assume a working machine-readable minimal association of keywords and vocabularies

The topic of creating a machine-readable vocabulary definition is closely related, but ***please*** let's keep that complex topic as separate as possible.  For the purpose of this discussion assume that:

1.  We will be able to create something that will at least indicate which keywords go with which vocabularies, most likely but not necessarily a separate file identified the vocabulary URI.
2.  We will be able to mark which of those keywords are SVAs in whatever format we come up with.

### Assume a strengthened dialect/vocabulary determination requirement

Currently it's conformant to ignore `$schema`, and the behavior of unknown `$schema` values or missing `$schema` is under-constrained.  There are numerous ideas on how to improve this, how to connect it to the media type registration, and how compatibility should be handled.  These ideas are already being discussed in _multiple_ other discussions, and should stay there.

Let's assume that whatever happens with all of that, in the future:

1. `$schema` or any replacement/addition in this area MUST be respected (including whatever behavior or range of allowable behaviors are defined for the absence of `$schema`), such that schemas are _always_ processed accordingly or are not processed at all (in a way that clearly informs the caller of the refusal to process).
2. Whatever this ends up looking like, it still has something analogous to setting vocabularies to `true` or `false` in `$vocabulary`, whether those exact keywords and values are still in use or not.  This discussion will simply refer to _required_ vs _optional_ vocabularies, which are currently implemented by `true` and `false` respectively.

### Assume better explanation of optional vocabularies

As of 2020-12, optional vocabularies exist for two purposes:

1.  to allow for the common case of vocabularies consisting of only SVAs to be implemented without custom code
2. to allow for vocabularies consisting of only non-validation-impacting keywords (so, SVAs but also templated annotations such as Hyper-Schema's `links`) to be gracefully ignored in validation-only contexts 
 
This is not as clear as it could be in the current specification, in part because in 2019-09 I had some other ideas in mind as shown by the weird handling of the format vocabulary.  All of that was (correctly) jettisoned in 2020-12.

Let's assume that we agree on use case 1, and reach a consensus on whether use case 2 and/or any other use cases are intended, and clarify that in the spec.  SVA-only vocabularies are central to the question of removing non-vocabulary keywords, so they are the main thing to worry about here. 

## Problem

Vocabularies carve out namespaces allowing keywords of the same name from different sources to be distinguished.  Schema authors can avoid conflicts among identically named keywords by managing namespaces through `$vocabulary`.

Allowing non-vocabulary keywords means that the entire possible set of keyword names is a single namespace that is always available.  While a meta-schema can describe non-vocabulary keywords, there are no normative requirements regarding enabling or disabling them.  Implementations can enable or disable them at will, including by default, potentially masking or contradicting vocabulary keywords.

This means that any new keyword added to a vocabulary (including the JSON Schema core vocabulary) might conflict with an always-enabled (or unpredictably-enabled) non-vocabulary keyword in the wild.  Therefore, no keyword addition can be considered safe, **_making forward compatibility impossible to guarantee_**.

This has also led to some people assuming an ability to treat standard keywords as non-vocabulary keywords as a way to ignore `$vocabulary` controls, which directly violates schema/meta-schema author intent.

### Keyword recognition

Since there is no mechanism to associate keywords and vocabularies, there are currently a two cases that an implementation might encounter:

* supported keywords from vocabularies in `$vocabulary` (either `true` or `false`)
* unknown keywords

Currently, if an optional vocabulary is unsupported, its keywords appear unknown in the same way as non-vocabulary keywords.

### Vocabulary creation

Creating a vocabulary right now requires

* specifying the keywords
* assigning a URI to the vocabulary
* creating a vocabulary meta-schema and assigning a URI to it
* _assuming we add a file for associating keywords with vocabularies, one would need to write that as well_

Using the vocabulary requires writing a custom meta-schema that adds the vocabulary URI and references the associated meta-schema, and referencing that custom meta-schema in the schema.

All of this has to be done even if the vocabulary is only used as an optional vocabulary and therefore does not require custom code.

## The proposal

1. Non-vocabulary keywords are banned, meaning that they cause an error when encountered even if the meta-schema allows them
2. Since the vocabulary-keyword association file notes which keywords are SVAs, an SVA-only vocabulary can automatically be detected
3. SVA-only vocabularies are considered supported regardless of whether they are optional or required, as they do not require custom code to implement
4. Since there are no longer unknown keywords, there is no need to specify what to do with them
5. SVA-keywords MUST always be collected as annotations (assuming annotations are being collected) no matter whether the vocabulary that defines them is optional-supported, optional-unsupported, or required
6. Non-SVA keywords from optional vocabularies SHOULD be collected as annotations, and MUST otherwise be ignored

## Tradeoffs

In doing this, we lose the ability to casually use additional keywords without declaring a meta-schema and vocabulary for using them.  Let's say we had a keyword `whatever` that we just wanted to have treated as an annotation.  Currently, you can just use it without doing anything else.  

This proposal would require:

* assigning a URI to a vocabulary containing `whatever` (we can emphasize that there are very easy ways to assign URIs such as `tag:your-email@example.com,2022:your-vocab-name` so people don't need to worry about hostnames, HTTPS, etc.)
* writing the keyword-vocabulary association and providing it somehow
* writing a custom meta-schema including the vocabulary and providing it somehow
* referencing the custom meta-schema

Note that while you _can_ create a meta-schema for your vocabulary and reference it, this is not strictly required.

Various ideas that folks have had (_to be discussed elsewhere!_) for inlining keywords in vocabulary declarations, or even inlining extension meta-schemas in schemas, could reduce the impact of this scenario (and vocabularies in general, which is why the overall topic of streamlining vocabularies belongs in another discussion).

This could be as low-impact as associating a URI with a list of keyword names (and SVA-ness) and writing something slightly more complex than `$schema` with a URI in the schema to indicate the usage.

I think it's worth doing, and also it points to several other topics that we would need to prioritize in order to make it viable.

The current approach is too heavy-weight, but for _specifically_ replacing the "collect unknown keywords as annotations" case, I think we can make it work in a more lightweight manner.

**Comment by gregsdennis** [comment]
We can use the "inline" suggestions to enable an ad-hoc vocab that's only valid within the meta-schema that declares it.  This would eliminate the need for a URI and other boilerplate things.

This may only be a valid approach for SVA vocabs, though as non-SVAs would need code.

**Reply by handrews** [reply]
Yeah that's a good possible way to streamline.  SVA-only vocabs are my main concern here, as they are the only behavior that can currently be accomplished without code anyway.  We need to preserve that in a reasonably lightweight manner, but we don't need to improve any of the other cases (in this discussion, anyway).

**Comment by jdesrosiers** [comment]
> 6. Non-SVA keywords from optional vocabularies SHOULD be collected as annotations, and MUST otherwise be ignored

I assume this should say, "Non-SVA keywords from optional vocabularies **_that are not supported by the implementation_** SHOULD be collected as annotations".

I think this could be problematic. If we collect the value of a non-SVA keyword as an annotation, the annotation result specified by the vocabulary might be different than the one collected. A consumer of the annotations may expect one or the other annotation result and might end up broken if the validator adds or removes support for that vocabulary.

I think the safest thing to do for a non-SVA keyword from an optional vocabulary that isn't supported is to completely ignore it, including no output unit for that keyword in the results.

**Reply by handrews** [reply]
@jdesrosiers yes to the "that are not supported by the implementation" part.

What you propose would be a change from the current behavior of collecting every such keyword as an annotation regardless of its nature.  Is that intentional?

What you are proposing would effectively create two distinct (rather than overlapping) use cases for optional vocabularies:

* SVA-only vocabularies, which can be supported without code as optional vocabularies.  These should basically always be considered "optional", which raises the question of whether the optional/required distinction even matters- if an implementation can determine that a vocabulary is SVA-only, it can support it as a required vocabulary as that's all it needs to "understand"
* non-validation-impacting, non-SVA-only vocabularies, which would just be ignored.  This is the "please don't refuse to process my hyper-schema just because you don't understand hyper-schema" scenario.  The pro of what you suggest is that if the caller gets the `links` annotations at all, they will be filled out properly.  The con is that if the implementation doesn't support hyper-schema, the caller gets nothing and does not even get the benefit of knowing which links are valid for the instance, even if they have to do all of the template-filling themselves.

A middle path for the second case would involve indicating in the output that the annotations were collected blindly, at which point the caller can determine if they want to bother.  Variations on this would allow the caller to signal whether they want blind collection, or whether they want a refusal, which would be a way of balancing/negotiating schema author and caller requirements.

**Reply by jdesrosiers** [reply]
> What you propose would be a change from the current behavior of collecting every such keyword as an annotation regardless of its nature. Is that intentional?

Yes, it's intentional. Your proposal didn't create the problem I identified, but if we're fixing the problems with unknown keyword annotation collection, why leave some of the problems unfixed?

I agree with all the rest of what you wrote. The only way I see it being safe to collect non-SVA unimplemented keywords as SVAs is if they are distinguishable from the fully implemented keyword result. For example, in my implementation, every keyword has a URI. Let's say that the URI for `links` is `https://json-schema.org/keywords/links`. If the `links` were collected as an SVA it could be identified as `https://json-schema.org/keywords/unknown#links`. Someone could code against the "unknown-links" URI and the "links" URI and seamlessly switch over when the validator suddenly supports the hyper-schema vocabulary.

**Reply by handrews** [reply]
@jdesrosiers 

> Yes, it's intentional. Your proposal didn't create the problem I identified, but if we're fixing the problems with unknown keyword annotation collection, why leave some of the problems unfixed?

Yeah, you've identified a really great opportunity and solution here!

And yes, any mechanism that allows someone to determine what they're getting would allow for a reasonable grey zone.  If combined with the ability to request specific annotations (#236), that would make it easier to determine if an evaluation is safe or not.  Potentially you could request that non-SVAs only be collected if fully supported, or request them to be collected regardless.

**Comment by handrews** [comment]
Several of this proposal's assumptions are now being tracked or addressed directly:

* Better explanation of optional vocabularies is addressed by PR json-schema-org/json-schema-spec#1295
* Strengthening the requirements around optional vocabularies is being tracked as issue json-schema-org/json-schema-spec#1300
* Strengthening the requirements around dialect/vocabulary determination is being tracked as issue json-schema-org/json-schema-spec#1301

**Comment by gregsdennis** [comment]
Related: https://github.com/json-schema-org/json-schema-spec/issues/1365 and associated PR https://github.com/json-schema-org/json-schema-spec/pull/1244 (@awwright).

**Comment by gregsdennis** [comment]
I expect a consequence of not supporting unknown keywords is no longer being able to process schemas with unsupported optional vocabularies (which... why then have the ability to declare them optional?), unless the vocabulary can be expressed in a meaningful machine-readable way that says, "These keywords are mine," so that they implementation can know to ignore those (because the vocab is optional).

---

_For another discussion perhaps_

Are optional vocabularies even useful?  One of the goals we have is interoperability.  If a schema that uses an optional vocab is processed by two implementations, one of which understand that vocab and the other doesn't, then they could have differing validation results.  (Maybe the meta-schema author included a vocab with validation keywords as optional.)  This breaks interoperability.

Is that something we need to cover since it was a user that wrote that meta-schema?

**Reply by SorinGFS** [reply]
> Are optional vocabularies even useful?

Very good question! 

> One of the goals we have is interoperability. If a schema that uses an optional vocab is processed by two implementations, one of which understand that vocab and the other doesn't, then they could have differing validation results. (Maybe the meta-schema author included a vocab with validation keywords as optional.) This breaks interoperability.

Totally agree!

Introducing vocabulary system IMHO was probably one of the main reasons for people to choose not to upgrade. The vocabulary system not only introduces a high level of confusion, but as you say it breaks interoperability, in other words it damages exactly the main purpose of this project, namely to ensure a common language for data transport.

Moreover, the introduction of vocabulary system forces a chain of subsequent repairs of some problems that derive from its introduction.

**Reply by Relequestual** [reply]
I imagine the simplest approach to disallowing unknown keywords is to collect them by analysis of vocabulary meta-schemas.
Additionally, a dialect schema should `$ref` all vocabularies meta-schemas, and a dialect meta-schema would then just need to set `unevaulatedProperties: false`.

The use case for optional vocabularies is actually still strong, and when I explain it to people (including specific orgs in the industry) in the wild, they actually want it.

The use case is this: 
Imagine you use JSON Schema as your single source of truth for data models.
Now, you want to generate your front end forms and your back end databases using the same schema, and you want the same schema to sit on your API gateway to do validation.

You want to define all of those things in the same schema, but without it causing a problem for validaiton implementaions that don't know how to generate a database and don't know how to generate forms.

The tooling that DOES know how to do that things, will do those things it's expected to do, having recognised the vocabularies.

**Reply by SorinGFS** [reply]
> You want to define all of those things in the same schema, but without it causing a problem for validaiton implementaions that don't know how to generate a database and don't know how to generate forms.

Allowing unknown keywords by default wouldn't have the same impact?

**Reply by jdesrosiers** [reply]
> I expect a consequence of not supporting unknown keywords is no longer being able to process schemas with unsupported optional vocabularies

Right. I brought that up somewhere before. An implementation doesn't need to support the keywords in the optional vocabularies, but it does need to know what keywords the vocabulary defines. Currently, that requires code, but we've discussed some options for a vocabulary description document that would solve a problem like this. Another solution is doing something similar to what Austin has proposed and say that any keyword that starts with a `@` is a simple value annotation with no validation assertions. That would be useful for people who simply want to make random annotations that don't need to be part of a vocabulary.

**Reply by jdesrosiers** [reply]
> Are optional vocabularies even useful?

Ben mentions annotation-only vocabularies. It's less useful to have and optional vocabulary that does validation, but it can be useful in cases where a partial validation result is better than no validation at all. The example I think of is a schema that an API provides for a request. An API client can use the schema to check if the request is likely to be accepted by the API. The API uses the schema to verify if it should accept the schema. Let's say the server-side uses a custom vocabulary, such as `data`, to do some extra checks standard JSON Schema can't do. Not all clients are going to understand the `data` keyword. If `data` is optional, clients can get most of the validation checks which leaves them better off than if `data` was required and they got no validation at all.

**Reply by SorinGFS** [reply]
@jdesrosiers 
Continuing to support unknown keywords by default wouldn't be enough to allow you to completly eliminate vocabulary system?

**Reply by gregsdennis** [reply]
@SorinGFS your comments across the various discussions recently show that you have a different vision for this project than the majority of those who build it, implement it, use it, and provide feedback on it.

If you have commentary that helps move this project toward the collective vision of those groups of people, by all means, please share.

However attempting to break down the core ideas and concepts of this project in order to turn it into something completely different is not on the table for discussion.

If you have questions regarding _why_ JSON Schema is built as it is, we are happy to discuss those with you (elsewhere), and hopefully help your understanding to align with our vision.  But please understand that injecting those ideas into these discussions is not appropriate.  Please join us in Slack instead.

These particular discussions are for finalizing the details of general decisions we've already agreed upon, not for completely overhauling JSON Schema.

I will be marking these comments as unconstructive or unrelated when I have the next opportunity.

Thank you.

**Reply by SorinGFS** [reply]
Ok, np.

**Reply by benjagm** [reply]
Thanks to @gregsdennis @jdesrosiers @handrews @Relequestual @SorinGFS and everyone in this conversation for sharing their views. As all of you know, Evolve JSON Schema requires a lot of deep discussions to reach the best project decision, which demands a lot of time, energy, and focus.... so huge THANKS!

@SorinGFS you brought great comments and a different vision that is refreshing THANKS!. I think it will be super interesting for the Community if you can start a [new discussion topic](https://github.com/json-schema-org/community/discussions/new/choose) in the Community sharing your vision so it can be canalized in a way that brings the maximum value. And of course, will be great to see you in [Slack](https://json-schema.org/slack) too.

**Reply by SorinGFS** [reply]
@gregsdennis 
I don't want to polute this topic but I need to respond, please feel free to hide or remove. Also, please don't be offended by my opinions, despite the differences I really appreciate your work because I have the same concerns and I know how hard it is to fit all the details in the head at the same time.
@benjagm 
Thanks for the appreciation, but I have no reason to continue for the moment. Things are quite clear about the direction in which it is going and it is radically opposite to what I think. It's unknown keywords vs vocabulary, one must go, and not the one I want. It's about more flexibility and simplicity vs more strictness and complexity. There is really no point in opening another discussion topic because in the context of the changes that have been decided, I really have no ideas. All that remains is to see what comes out and possibly we can continue from there.

**Comment by SorinGFS** [comment]
@gregsdennis 
...continuation from inappropriate [topic](https://github.com/orgs/json-schema-org/discussions/295#discussioncomment-4834427).

> Would you still have this stance if adequate tooling were provided to aid in the migration? For example, if [alterschema](https://github.com/sourcemeta/alterschema) were to support migrating any unknown keywords to one of the proposed solutions:
>    [@foo convention](https://github.com/orgs/json-schema-org/discussions/241#discussioncomment-4801155)
>    [ad-hoc annotation vocabs](https://github.com/orgs/json-schema-org/discussions/241#discussioncomment-3672231)
> (Again, I'd like to have proposals discussed over there. Let's keep this to the "breaking changes" topic.)

I'm aware of `alterschema` existance, for my small projects I don't think it's relevant, I can easily make the transition to anything. I think that those with complicated projects should be asked what their opinion is.

**Reply by SorinGFS** [reply]
@gregsdennis 
> Would you still have this stance if adequate tooling were provided to aid in the migration? For example, if [alterschema](https://github.com/sourcemeta/alterschema) were to support migrating any unknown keywords to one of the proposed solutions:
[@foo convention](https://github.com/orgs/json-schema-org/discussions/241#discussioncomment-4801155)
[ad-hoc annotation vocabs](https://github.com/orgs/json-schema-org/discussions/241#discussioncomment-3672231)

First proposed solution means a simple parse and replace in the schema, then same parse and replace in depending components. 
Second proposed solution would be much more complicated than first one, because keywords are to be found in a different object.
Second proposed solution would be cleaner than first one, but would require more work.
IMO if you want a competent opinion, I think that the tool you are talking about should first exist and be capable of conversions that can then be tested by those interested for several scenarios. As far as I know, there is not even an example that can be analyzed for the second scenario. If there was now a tool that would work, maybe those interested would imagine more easily what they have to do and maybe they would be favorable to the change...

**Comment by westurner** [comment]
> I expect a consequence of not supporting unknown keywords is no longer being able to process schemas with unsupported optional vocabularies (which... why then have the ability to declare them optional?), unless the vocabulary can be expressed in a meaningful machine-readable way that says, "These keywords are mine," so that they implementation can know to ignore those (because the vocab is optional).

Would the proposed change break Linked Data and JSON-LD (and json-ld-schema), which use namespaced:URIs for some or all attributes?

A jsonschema parser could raise a warning when there are unknown keywords in the schema, but I suspect that excluding unknown keywords would limit reusability of jsonschema specifically for linked data use cases. There exist validators more complex than jsonschema for which there hopefully needn't be yet another schema definition document, that unreserved:keywords could be used in to define *online* validation  with for example.

Linked Data use cases to make sure this change wouldn't break, that should be added to the Use Cases document:

- validate schema.org JSONLD, where attributes are full URIs or uri:CURIEs

- CURIE: https://en.wikipedia.org/wiki/CURIE
- (W3C) QName: https://en.wikipedia.org/wiki/QName

**Reply by jdesrosiers** [reply]
JSON Schema has no relation to Linked Data, JSON-LD, and Schema.org. Nothing we do will effect your use of those technologies.

**Reply by westurner** [reply]
> JSON Schema has no relation to Linked Data, JSON-LD, and Schema.org. Nothing we do will effect your use of those technologies.

Is that confirmed with test cases and/or a IDK a statement of intent to work well with Linked Data, too?

**Reply by Relequestual** [reply]
> > JSON Schema has no relation to Linked Data, JSON-LD, and Schema.org. Nothing we do will effect your use of those technologies.
> 
> Is that confirmed with test cases and/or a IDK a statement of intent to work well with Linked Data, too?

I don't really know what this means. There is nothing in the JSON Schema specifcation that links to JSON-LD or Schema.org.
In what way do you see them as interacting?

**Reply by westurner** [reply]
JSON Schema MAY be used to validate data that is also already JSON-LD or that is also validated with SHACL.

How would such a statement of intent to not break Linked Data - e.g. the aforementioned specs - affect whether for example the existing json-ld-schema test cases which may not (?) even currently use custom keywords or extension vocabularies?: 
https://github.com/mulesoft-labs/json-ld-schema/tree/master/src/test/data/schema

- [ ] JSON Schema test case: Schema.org IDK ScholarlyArticle as JSON-LD
  - and then also with SHACL but which repo?

**Reply by Relequestual** [reply]
>JSON Schema MAY be used to validate data...

I think that's kind of the point here. JSON Schema validates data. JSON data. JSON Schema doesn't care what sort of JSON data, as long as it's valid JSON data.

Any custom keywords or extensions can be formed into a vocabulary (or something else should we provide another method for pure embedding or signalling pure annotation extensions).

The fact that the data is JSON-LD, or any other data structure that uses JSON, matters not. Unless I'm really missing something here, in which case, please enlighten me.

----

Taking a look at the specific example you linked, I can see there's a custom dialect and vocabulary. That's great to see! It uses JSON Schema 2019-09. Updating it to the "non-breaking era" JSON Schema looks like it would be trivial. As for tooling? It really depends how it's being used, so I couldn't comment.

**Comment by SorinGFS** [comment]
@gregsdennis 

Couldn't json-schema decentralization help avoiding to disallow non-vocabulary keywords breaking change? For example, [@cfworker/json-schema](https://github.com/cfworker/cfworker/blob/main/packages/json-schema/README.md) library treats each keyword in each draft as a separate schema. In this way, this library finds the correct interpreter for any combination of schemas.

**Reply by gregsdennis** [reply]
> json-schema decentralization

I don't understand what you mean by this or how it relates to disallowing non-vocab keywords?

> For example, [@cfworker/json-schema](https://github.com/cfworker/cfworker/blob/main/packages/json-schema/README.md) library treats each keyword in each draft as a separate schema.

Do you mean it "treats each keyword in each draft as a separate _keyword_?"  A keyword by itself isn't a schema.

If so, users typically expect that a keyword in one version means the same thing and operate the same way in the next version.  Their primary complaint is that keywords keep changing from one version to the next.

**Reply by SorinGFS** [reply]
> Do you mean it "treats each keyword in each draft as a separate keyword?" A keyword by itself isn't a schema.

Agree, a keyword is not a schema, but behind a keyword is an intention which can this way to be identified and coupled with the right interpreter, like a translator. 

> users typically expect that a keyword in one version means the same thing and operate the same way in the next version. Their primary complaint is that keywords keep changing from one version to the next.

The point is if you have a way to identify back in time which intent was behind an used keyword it doesn't matter how it changed. You will be able to colect and map all the intents to their correct actions. As an analogy, that is how `pandoc` works and is able to convert almost any document format into any document format. They do not care how a document marks its formats, they have their own engine which is non human readable in which they map every existing intent in a document, then they translate all the intents to another document type using their formatting rules. See this [picture](https://pandoc.org/diagram.svgz?v=20230203095535). 

This technique will bring backwards compatibility, will provide to users a way for easy transition to latest release, there will be no reason for them to stick to older drafts. Plus, in next draft or drafts if you just add new kewords and don't change older ones I think that will satisfy everybody.

**Reply by Relequestual** [reply]
I believe I understand what you're saying, and it is something that has been discussed previously, and may come up again, however...

The pandoc anaolgy doesn't hold. Pandoc is a tool for conversion. It exists and can be used when the need arrises. JSON Schema is a standard, with many implementations across many languages. The schema itself is supposed to be interoperable across them all, providing the same validation result. They cannot all be expected to also provide previous to current translation, nor bolt on a specific tool in a specific language.

While there is a tool which does schema upgrades, it is stand alone. Making it required that people implement it to be a compliant validation is unreasonable.

**Reply by SorinGFS** [reply]
> While there is a tool which does schema upgrades, it is stand alone. Making it required that people implement it to be a compliant validation is unreasonable.

I also think the same, it shouldn't be required to use a tool. And I imagine this working based only on an instruction set about re-referencing old drafts and old-draft based schemas in a decentralized structure, and the next drafts being released directly in that decentralized format. The users of new drafts would not even need to be aware of that set of instructions. In the next period I will try to provide a functional example.

**Reply by SorinGFS** [reply]
@Relequestual 

As I promised, here is my view of [decentralized json-schema](https://github.com/SorinGFS/decentralized-json-schema). The work is still at the beginning, a lot of work to do identify all the rules that must apply, but at least you can have a view of what I try to achieve. For exemple, by using it I was able to detect all the keywords used since the first draft:


<details>
<summary>
Detected keyword definitions across versions from draft-00 to draft/next:
</summary>

```shell
============================================
= = = = = = = 1 1 1 $anchor
= = = = = = 1 1 1 1 $comment
= = = = = = = = 1 1 $dynamicAnchor
= = = = = 1 1 1 1 1 $id
= = = = = = = 1 1 1 $recursiveAnchor
1 1 1 1 1 1 1 1 1 1 $ref
1 1 1 1 1 1 1 1 1 1 $schema
= = = = = = = 1 1 1 $vocabulary
= = = = = = = 1 1 = absoluteKeywordLocation
= = = 1 1 1 1 1 = = additionalItems
1 1 1 1 1 1 1 1 1 1 additionalProperties
= = = = 1 1 1 1 1 1 allOf
1 1 1 = = = = = = = alternate
= = = = = = 1 1 1 = anchor
= = = = = = 1 1 1 = anchorPointer
= = = = = = = 1 1 1 annotations
= = = = 1 1 1 1 1 1 anyOf
= = = = = = 1 1 = = attachmentPointer
= = = = = 1 1 1 = = base
= = = = 1 1 = = = = binaryEncoding
= = = = = 1 1 1 1 1 const
= = = = = 1 1 1 1 1 contains
1 1 1 1 = = 1 1 1 1 contentEncoding
= = = = = = 1 1 1 1 contentMediaType
= = = = = = = 1 1 1 contentSchema
= = = = = = 1 1 = = contextPointer
= = = = = = 1 1 = = contextUri
1 1 1 1 1 1 1 1 1 1 default
= = = 1 1 1 1 1 1 1 dependencies
= = = = = = = 1 1 1 dependentRequired
= = = = = = = 1 1 1 dependentSchemas
= = = = = = = 1 1 1 deprecated
1 1 1 1 1 1 1 1 1 1 description
1 1 1 1 = = = = = = disallow
= = 1 1 = = = = = = divisibleBy
= = = = = = = = = 1 droppedAnnotations
= = = = = = 1 1 1 1 else
= = = = 1 = = = = = encType
1 1 1 1 = = = = = = enctype
1 1 1 1 1 1 1 1 1 1 enum
= = = = = = = = 1 = error
= = = = = = = 1 1 1 errors
= = = = = = = = = 1 evaluationPath
= = = = = 1 1 1 1 1 examples
= = = 1 1 1 1 1 1 1 exclusiveMaximum
= = = 1 1 1 1 1 1 1 exclusiveMinimum
1 1 1 1 = = = = = = extends
1 1 1 1 1 1 1 1 1 1 format
1 1 1 1 1 = = = = = fragmentResolution
= = = = = = 1 1 1 = headerSchema
1 1 1 1 1 1 1 1 1 = href
= = = = = = 1 1 = = hrefInputTemplates
= = = = = = 1 1 = = hrefPrepopulatedInput
= = = = = 1 1 1 1 = hrefSchema
= = = = = = = 1 = = https://json-schema.org/draft/2019-09/vocab/applicator
= = = = = = = 1 = = https://json-schema.org/draft/2019-09/vocab/content
= = = = = = = 1 = = https://json-schema.org/draft/2019-09/vocab/core
= = = = = = = 1 = = https://json-schema.org/draft/2019-09/vocab/format
= = = = = = = 1 1 = https://json-schema.org/draft/2019-09/vocab/hyper-schema
= = = = = = = 1 = = https://json-schema.org/draft/2019-09/vocab/meta-data
= = = = = = = 1 = = https://json-schema.org/draft/2019-09/vocab/validation
= = = = = = = = 1 = https://json-schema.org/draft/2020-12/vocab/applicator
= = = = = = = = 1 = https://json-schema.org/draft/2020-12/vocab/content
= = = = = = = = 1 = https://json-schema.org/draft/2020-12/vocab/core
= = = = = = = = 1 = https://json-schema.org/draft/2020-12/vocab/format-annotation
= = = = = = = = 1 = https://json-schema.org/draft/2020-12/vocab/format-assertion
= = = = = = = = 1 = https://json-schema.org/draft/2020-12/vocab/meta-data
= = = = = = = = 1 = https://json-schema.org/draft/2020-12/vocab/unevaluated
= = = = = = = = 1 = https://json-schema.org/draft/2020-12/vocab/validation
= = = = = = = = = 1 https://json-schema.org/draft/next/vocab/applicator
= = = = = = = = = 1 https://json-schema.org/draft/next/vocab/content
= = = = = = = = = 1 https://json-schema.org/draft/next/vocab/core
= = = = = = = = = 1 https://json-schema.org/draft/next/vocab/format-annotation
= = = = = = = = = 1 https://json-schema.org/draft/next/vocab/format-assertion
= = = = = = = = = 1 https://json-schema.org/draft/next/vocab/meta-data
= = = = = = = = = 1 https://json-schema.org/draft/next/vocab/unevaluated
= = = = = = = = = 1 https://json-schema.org/draft/next/vocab/validation
1 1 1 1 1 = = = = = id
= = = = = = 1 1 1 1 if
= = = = = = = 1 1 1 instanceLocation
1 1 1 1 1 1 1 1 1 1 items
= = = = = = = 1 1 = keywordLocation
1 1 1 1 1 1 1 1 1 = links
= = = = = = = 1 1 1 maxContains
1 1 = = = = = = = = maxDecimal
1 1 1 1 1 1 1 1 1 1 maxItems
1 1 1 1 1 1 1 1 1 1 maxLength
= = = = 1 1 1 1 1 1 maxProperties
1 1 1 1 1 1 1 1 1 1 maximum
1 1 1 = = = = = = = maximumCanEqual
= = = = 1 1 = = = = media
1 1 1 1 1 1 = = = = mediaType
1 1 1 1 1 = = = = = method
= = = = = = = 1 1 1 minContains
1 1 1 1 1 1 1 1 1 1 minItems
1 1 1 1 1 1 1 1 1 1 minLength
= = = = 1 1 1 1 1 1 minProperties
1 1 1 1 1 1 1 1 1 1 minimum
1 1 1 = = = = = = = minimumCanEqual
= = = = 1 1 1 1 1 1 multipleOf
= = = = = = = = = 1 nested
= = = = 1 1 1 1 1 1 not
= = = = 1 1 1 1 1 1 oneOf
1 1 1 = = = = = = = optional
1 1 1 1 1 = = = = = pathStart
1 1 1 1 1 1 1 1 1 1 pattern
= = = 1 1 1 1 1 1 1 patternProperties
= = = = = = = = 1 1 prefixItems
1 1 1 1 1 1 1 1 1 1 properties
= = = = = = = = = 1 propertyDependencies
= = = = = 1 1 1 1 1 propertyNames
= = = = = 1 1 1 1 1 readOnly
1 1 1 1 = = = = = = readonly
1 1 1 1 1 1 1 1 1 = rel
= = = 1 1 1 1 1 1 1 required
1 1 1 1 = = = = = = requires
1 1 1 1 = = = = = = root
= = = = 1 = = = = = schema
= = = = = = = = = 1 schemaLocation
= = = = = 1 = = = = submissionEncType
= = = = = = 1 1 1 = submissionMediaType
= = = = = 1 1 1 1 = submissionSchema
= = = = = = 1 1 1 = targetHints
= = = = = = 1 1 1 = targetMediaType
= = 1 1 1 1 1 1 1 = targetSchema
= = = = = = 1 1 = = targetUri
= = = = = = 1 1 1 = templatePointers
= = = = = = 1 1 1 = templateRequired
= = = = = = 1 1 1 1 then
1 1 1 1 1 1 1 1 1 1 title
1 1 1 1 1 1 1 1 1 1 type
= = = = = = = 1 1 1 unevaluatedItems
= = = = = = = 1 1 1 unevaluatedProperties
= = 1 1 1 1 1 1 1 1 uniqueItems
= = = = = = = 1 1 1 valid
= = = = = = 1 1 1 1 writeOnly
============================================

all keywords count: 136
keyword definition referrers: 1062
keyword definition footprints: 467

======= checking invalid references =======
[ 'https://json-schema.org/draft/2020-12/meta/hyper-schema' ]
============================================
invalid references count: 1
```

</details>

**Comment by automaton82** [comment]
This feels like the same as XML's [Unique Particle Attribution](https://en.wikipedia.org/wiki/Unique_Particle_Attribution), discussion here:

https://lists.w3.org/Archives/Public/www-tag/2004Aug/att-0010/NRMVersioningProposal.html

Has that been reviewed? Does JSON schema want to make JSON into XML? Food for thought.

**Reply by gregsdennis** [reply]
Thanks for raising this.  That looks like a namespacing disambiguation.  That's only marginally related to what we're doing here.

Here, we just want to ensure that every keyword used in a schema is defined _somewhere_.  For us, that means in a vocabulary.  We don't have a concept of namespacing (though separately, we have been investigating how that could work).

> Does JSON schema want to make JSON into XML?

I don't see how we could make JSON into XML.  They're quite different syntaxes and data models.  XML does have a lot that's good, which is why it was king for a long time, so what would be wrong with wanting to support some of those features if we can?

**Comment by jherico** [comment]
HTTP and MIME both solved this issue with headers by allowing non-standard headers as long as they were prefixed by X-.  Seems like a similar solution would be possible here.  Even thought it would still break existing schema, it would also continue to allow organizations to create custom tooling with additional fields that aren't part of the declared vocabulary.  Failing to support some method for orgs to create additional fields will simply lead to them turning off schema validation in some cases.

**Reply by AjaxGb** [reply]
Agreed, reserving some prefix like `x-` or `_` for custom fields would be a simple and clean way to solve the issue.

**Reply by gregsdennis** [reply]
Please see https://github.com/orgs/json-schema-org/discussions/329#discussioncomment-4988859.

Also, prefixed headers like that are now advised against.
